src/lib/address_resolve/AddressResolve.h

```diff
@@ -0,0 +1,178 @@
+    /// or different interfaces will return separate 'done resolving' calls.
+    ///
+    /// If the min lookup time is set to 0, implementations are expected to call
+    /// 'OnNodeAddressResolved' as soon as the first DNSSD response is received.
+    NodeLookupRequest & SetMinLookupTime(System::Clock::Milliseconds32 value)
+    {
+        mMinLookupTimeMs = value;
+        return *this;
+    }
+
+    /// The maximum lookup time is how much to wait until a TIMEOUT error is
+    /// declared.
+    ///
+    /// If a DNSSD response is received before this max timeout, then
+    /// OnNodeAddressResolved will be called on listener objects (immediately)
+    /// if the first DNSSD reply arrives after MinLookupTimeMs has expired.
+    NodeLookupRequest & SetMaxLookupTime(System::Clock::Milliseconds32 value)
+    {
+        mMaxLookupTimeMs = value;
+        return *this;
+    }
+
+private:
+    static constexpr System::Clock::Milliseconds32 kMinLookupTimeMsDefault{ 200 };
+    static constexpr System::Clock::Milliseconds32 kMaxLookupTimeMsDefault{ 3000 };
+
+    PeerId mPeerId;
+    System::Clock::Milliseconds32 mMinLookupTimeMs = kMinLookupTimeMsDefault;
+    System::Clock::Milliseconds32 mMaxLookupTimeMs = kMaxLookupTimeMsDefault;
+};
+
+/// These things are expected to be defined by the implementation header.
+namespace Impl {
+class NodeLookupHandle; // MUST derive from NodeLookupHandleBase
+}
+
+class Resolver
+{
+public:
+    virtual ~Resolver() = default;
+
+    /// Expected to be called exactly once before the resolver is ever
+    /// used.
+    virtual CHIP_ERROR Init(System::Layer * systemLayer) = 0;
+
+    /// Initiate a node lookup for a particular node and use the specified
+    /// Lookup handle to keep track of node resolution
+    ///
+    /// If this returns CHIP_NO_ERROR, the following is expected:
+    ///   - exactly one of the listener OnNodeAddressResolved
+    ///     or OnNodeAddressResolutionFailed will be called at a later time
+    ///   - handle must NOT be destroyed while the lookup is in progress (it
+    ///     is part of an internal 'lookup list')
+    ///   - handle must NOT be reused (the lookup is done on a per-node basis
+    ///     and maintains lookup data internally while the operation is still
+    ///     in progress)
+    virtual CHIP_ERROR LookupNode(const NodeLookupRequest & request, Impl::NodeLookupHandle & handle) = 0;
+
+    /// Expected to be provided by the implementation.
+    static Resolver & Instance();
+};
+
+} // namespace AddressResolve
+} // namespace chip
+
+// outside the open space, include the required platform headers for the
+// actual implementation.
+// Expectations of this include:
+//   - define the `Impl::NodeLookupHandle` derifing from NodeLookupHandle
+//   - corresponding CPP file should provide a valid Resolver::Instance()
+//     implementation
+#include CHIP_ADDRESS_RESOLVE_IMPL_INCLUDE_HEADER
```

Comment:
We we didn't need to care about the by-valueness of the handle, all of this would not be required

Reply:
This is mostly so I do not pay a memory pool cost or use heap. Everything else is refactorable.

Reply:
Closing this but  leaving the comment above for NodeLookupHandle as a discussion point about memory allocation and tradeoffs. I made the tradeoff of ugly include (Fully agree it is ugly) but gaining compile-time memory and type safety.